File: src/types/browser.ts
/**
 * Wire types for the in-app browser (TODO group J #61) — the FROZEN CONTRACT
 * between the Rust backend (`src-tauri/src/commands/browser.rs`) and the React
 * frontend (`src/lib/browser.ts`, `src/stores/browser.ts`,
 * `src/components/browser/`).
 *
 * The architecture this contract encodes is **not** the one Plan A first
 * proposed. Plan C proved that an in-pane embedded child webview is
 * structurally impossible under WebKitGTK (`set_bounds` is a no-op for a
 * GtkBox-packed child), so there is no `browser_view_*` surface, no native view
 * composited over the pane, and therefore no suppression refcount and no
 * live-view LRU. What exists instead is two things:
 *
 *  1. **Reader mode** — the backend fetches the page, sanitizes it with the very
 *     same `ammonia` pipeline the mail client uses, and hands back inert HTML.
 *     It renders in the ordinary webview inside a script-less `<iframe sandbox="">`,
 *     exactly as a mail body does, so a browser tab is an ordinary DOM pane with
 *     no compositing surprises.
 *  2. **A separate hardened live window** — a real engine, in its own OS window,
 *     spawned by the backend on an explicit user gesture. The frontend renders
 *     the *control* and the consent UI; it never renders the page.
 *
 * Three rules make this file load-bearing rather than decorative:
 *
 * 1. **No command in this contract takes a filesystem path.** Downloads cross
 *    the boundary only through `browser_download_decide`, which raises the
 *    native OS save dialog *inside Rust*. A page that controls the bytes, the
 *    filename and the MIME type still has no IPC verb that names a path.
 * 2. Field names are snake_case because they are serde-serialized Rust structs
 *    verbatim — the convention `src/types/mail.ts` already follows (no
 *    `#[serde(rename_all = "camelCase")]` on any mail struct). **Command
 *    arguments**, by contrast, are camelCase, because Tauri converts an
 *    `invoke` argument name to snake_case on the Rust side (see
 *    `src/lib/mail.ts`'s `{ accountId, folderId }`).
 * 3. `SecurityState` is rendered **verbatim**. The frontend computes no security
 *    conclusion of its own; a value it does not recognize degrades to the
 *    unknown/insecure reading, never to "secure".
 *
 * Design rationale: `docs/browser_plan_a.md` (surface/integration),
 * `docs/browser_plan_b.md` (threat model, reader pipeline, address bar rules),
 * `docs/browser_plan_c.md` (engine feasibility, and why the pane is not a view).
 */

/**
 * What the backend's navigation policy makes of a URL, **before** anything is
 * fetched. The address bar performs its own shape checks first, but this is the
 * gate: `allowed: false` means no fetch happens, whatever the field says.
 */
export interface UrlVerdict {
  allowed: boolean;
  /** Why it was refused, in words, when `allowed` is false. */
  reason?: string;
  /**
   * The URL as it should be *shown* — punycode decoded, userinfo stripped,
   * decimal/octal/hex IP literals normalized. Never re-derive this in TS by
   * string search; a display that disagrees with the parser is the attack.
   */
  display_url: string;
  /** Set when the host has an `xn--` label: the ASCII form, shown beside the
   *  Unicode one. Never the Unicode form alone. */
  punycode_warning?: string;
  /** Lowercased scheme, e.g. `"https"`. */
  scheme: string;
  /** Loopback / private / link-local — the intranet class, which the UI names
   *  in words rather than showing a bare address. */
  is_loopback: boolean;
}

/** How the connection is secured. Rendered verbatim; the frontend computes none
 *  of it. `"unknown"` is a real state (a page that never loaded), not a bug. */
export interface SecurityState {
  tls: "secure" | "insecure" | "unknown";
  scheme: string;
  /** The host as shown to the user, after IDNA normalization. */
  host_display: string;
  punycode_warning?: string;
  /** True while an OpenVPN tunnel is up: browser traffic goes through it, like
   *  every other program on the machine (`docs/browser_plan_b.md` §9.2). */
  vpn_active: boolean;
}

/**
 * One fetched-and-sanitized page. `html` has already been through the backend's
 * `ammonia` pipeline and carries no script, no remote reference and no `href` —
 * it renders in a `sandbox=""` iframe and nowhere else.
 */
export interface ReaderPage {
  /** What the user asked for. */
  requested_url: string;
  /** Where the fetch actually landed after redirects. These differ more often
   *  than people expect, and the difference is worth showing. */
  final_url: string;
  /** The display form of `final_url` (see {@link UrlVerdict.display_url}). */
  display_url: string;
  title: string;
  /** Sanitized in Rust before it ever crosses IPC. */
  html: string;
  security: SecurityState;
  /** The page hit a size/element cap and was cut short. */
  truncated: boolean;
  /** How many remote references (images, styles, fonts) were dropped. Drives an
   *  informational banner — there is no control to unblock them, because there
   *  is no proxy behind one. */
  blocked_remote_assets: number;
}

/** A live (real engine) window the backend owns. `label` is the Tauri window
 *  label; it is an opaque handle here and is never used to build a path. */
export interface LiveWindowRef {
  label: string;
  display_url: string;
}

/** Payload of the `browser:live-state` event. */
export interface LiveWindowState {
  label: string;
  display_url: string;
  title: string;
  security: SecurityState;
  loading: boolean;
}

/**
 * Payload of `browser:download-requested`: a download the engine wants to make,
 * quarantined in the backend and **refused by default**. Nothing reaches the
 * user's filesystem until {@link DownloadOutcome} says the user said yes in an
 * OS-native save dialog the backend raised.
 */
export interface DownloadRequest {
  download_id: string;
  /** Already sanitized by the backend (path separators, `..`, bidi controls and
   *  reserved device names removed). Still rendered as a plain text node. */
  file_name: string;
  mime_type?: string;
  size_bytes?: number;
  /** The sniffed bytes disagree with the name or the declared type — the single
   *  strongest signal a download is hostile. Shown as a persistent warning. */
  sniff_mismatch: boolean;
}

/** What `browser_download_decide` actually did. `saved: false` with no
 *  `file_name` means the user cancelled and nothing was written. */
export interface DownloadOutcome {
  saved: boolean;
  file_name?: string;
}

/**
 * Payload of `browser:blocked` — the navigation policy refused a URL.
 *
 * `reason` is a **machine token**, never a sentence: `app-origin`,
 * `scheme:file`, `redirect-to-link-local`, `loopback`… The wording lives in
 * `src/lib/i18n.ts` in five languages, reached through
 * {@link import("../lib/browser").reasonPhrase}. Rendering `reason` directly is
 * how `redirect-to-link-local` ends up on a user's screen; the canonical token
 * list is `services::web_safety::REASON_TOKENS` and
 * `src/__tests__/BrowserTripwire.test.ts` fails if the two sides drift.
 *
 * `window_label` names the live-page window a refusal happened in. Every emitter
 * of this event is a live window, so without it the frontend could only guess
 * which surface to blame — and the guess would wrongly wipe an unrelated reader
 * tab's page.
 */
export interface BlockedNavigation {
  display_url: string;
  reason: string;
  window_label?: string;
}

/** Payload of `browser:live-closed`. */
export interface LiveWindowClosed {
  label: string;
}

/**
 * What this build can actually do, **asked of the backend at runtime**.
 *
 * The frontend must never hardcode a platform check: Windows is refused in v1
 * (`docs/browser_plan_b.md` §5.2 — WebView2's permission default is *ask*, with
 * a dialog Eldrun did not write, the same call `services::sandbox` already makes
 * for Docker on Windows), but that is the backend's statement to make, and it
 * may change without this file changing. `platform_note` is the explanation to
 * show in place of the live-page control — a plain, non-alarming sentence.
 */
export interface BrowserCapabilities {
  live_windows_supported: boolean;
  reader_supported: boolean;
  platform_note?: string;
}

/** Where a clicked/typed URL should open (`settings.browser_link_target`). */
export type LinkOpenTarget = "external" | "in_app" | "ask";

File: src-tauri/src/schema/browser.rs
//! Serde structs for the in-app browser (TODO group J #61).
//!
//! Same convention as `schema::mail`: snake_case field names, `rename_all =
//! "lowercase"` on the small closed enums, and every type here serializes to
//! exactly the shape the frontend's typed wrappers declare.
//!
//! **What is deliberately absent from every type in this file: a path.** No
//! command in `commands::browser` accepts one and no result carries one. A
//! download reports the *display name* it was saved under and nothing else —
//! where it went is the OS save dialog's business, and telling the frontend
//! would make the next feature request "let the page pick it".

use serde::{Deserialize, Serialize};

/// What the transport is actually doing, as a word rather than an icon. Users
/// do not read padlocks.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum TlsState {
    /// https, certificate validated against the OS trust store.
    Secure,
    /// Plain http. There is no third "mixed" state in v1 because we have no API
    /// to observe passive mixed content (see `browser_engine`'s gap notes).
    Insecure,
    /// Not determined — a live window before its first load completes, or a
    /// scheme with no transport (`about:blank`).
    #[default]
    Unknown,
}

/// The security readout the chrome renders beside the address.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SecurityState {
    pub tls: TlsState,
    pub scheme: String,
    /// The host in its **Unicode** form, for display.
    pub host_display: String,
    /// The ASCII/punycode host, present only when it differs from
    /// `host_display`. The chrome must render **both** — the Unicode form alone
    /// is the homograph attack and the ASCII form alone is unreadable.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub punycode_warning: Option<String>,
    /// Whether an OpenVPN tunnel Eldrun knows about is up. Browser traffic goes
    /// through it exactly like every other process on the machine; this is here
    /// so that is *visible* rather than surprising.
    pub vpn_active: bool,
}

impl Default for SecurityState {
    fn default() -> Self {
        Self {
            tls: TlsState::Unknown,
            scheme: String::new(),
            host_display: String::new(),
            punycode_warning: None,
            vpn_active: false,
        }
    }
}

/// The result of the pure navigation-policy check.
///
/// Note the three-into-two mapping, because it is the one place this type is
/// subtler than it looks. The gate has **three** outcomes (allow / confirm /
/// block) and this struct has two fields:
///
/// | gate      | `allowed` | `reason`       |
/// |-----------|-----------|----------------|
/// | Allow     | `true`    | `None`         |
/// | Confirm   | `true`    | `Some(token)`  |
/// | Block     | `false`   | `Some(token)`  |
///
/// So `allowed && reason.is_some()` means *"reachable, but this is a loopback
/// or private-network address and the user should be told before you open
/// it"*. `reason` is a stable machine token (`loopback`, `private-network`,
/// `link-local`, `internal-name`, `app-origin`, `scheme:file`, `downgrade`,
/// `redirect-loop`, `unparsable`, `about-internal`, `no-host`) — the wording
/// lives in the frontend's i18n, never here.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct UrlVerdict {
    pub allowed: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
    /// The URL as it should be shown: userinfo removed, host in Unicode form,
    /// format controls stripped. **Never truncated** — eliding
    /// `https://example.com.evil.tld/…` to `https://example.com…` is the attack.
    pub display_url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub punycode_warning: Option<String>,
    pub scheme: String,
    pub is_loopback: bool,
}

/// One fetched-and-sanitized page. `html` is a **fragment** with no `href`, no
/// remote-loading attribute and no script anywhere — the frontend renders it in
/// an `<iframe sandbox="" srcdoc>` with its own `default-src 'none'` CSP, which
/// is a third layer nothing here relies on.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReaderPage {
    pub requested_url: String,
    /// Where the redirect chain actually ended. Not the same string as
    /// `requested_url` whenever a hop happened, and the difference is the whole
    /// point of reporting it.
    pub final_url: String,
    pub display_url: String,
    pub title: String,
    pub html: String,
    pub security: SecurityState,
    /// The body hit the element budget and was cut.
    pub truncated: bool,
    /// How many remote references the sanitizer dropped, for the "Eldrun
    /// blocked n remote images" banner.
    pub blocked_remote_assets: u32,
}

/// A live-page window. The label is opaque to the frontend and is the only
/// handle it ever gets — there is no window id, no native handle, no path.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LiveWindowRef {
    pub label: String,
    pub display_url: String,
}

/// `browser:live-state` — pushed whenever a live window loads, retitles, or
/// finishes.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LiveWindowState {
    pub label: String,
    pub display_url: String,
    /// Already stripped of bidi/format controls. A tab titled
    /// `example.com — Secure  \u{2069}` is a real technique.
    pub title: String,
    pub security: SecurityState,
    pub loading: bool,
}

/// `browser:live-closed`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LiveWindowClosed {
    pub label: String,
}

/// `browser:blocked` — a navigation the gate refused. A page state, not an
/// interrupt: the frontend renders an in-app error page, never a native modal.
///
/// `window_label` is what makes the event *attributable*. Every emitter of this
/// event today is a **live-page window** (its `on_navigation` gate, and its
/// download handler refusing an oversized transfer), which has nothing to do
/// with any reader tab — so without the label the frontend's only option was to
/// guess "the tab that most recently asked for a load", and that guess is wrong
/// in exactly the common case: a live window blocked mid-browse would wipe the
/// page an unrelated reader tab was showing. The reader's own refusals never
/// come through this event at all; they are the return value of
/// `browser_check_url`, which the caller already holds.
///
/// It stays `Option` rather than being made mandatory so a future emitter that
/// genuinely has no window (a background policy pass) can say so instead of
/// inventing a label.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct BlockedNavigation {
    pub display_url: String,
    pub reason: String,
    /// The `browser-*` label of the live window this happened in, when it
    /// happened in one.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub window_label: Option<String>,
}

/// `browser:download-requested` — bytes are in quarantine and nothing has
/// reached the user's filesystem. `file_name` is a **display name**, already
/// through `web_safety::sanitize_attachment_name`; it is not where anything is.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DownloadRequest {
    pub download_id: String,
    pub file_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub size_bytes: Option<u64>,
    /// The bytes are not what the name claims — either the sniffed type
    /// disagrees with the extension, or the content sniffs as an executable
    /// regardless of what it is called. A program download is **labelled, not
    /// refused**: people download installers, and a browser that cannot is
    /// broken.
    pub sniff_mismatch: bool,
}

/// What `browser_download_decide` did. Carries a name for the toast and
/// **never a path** — see the module note.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
pub struct DownloadOutcome {
    pub saved: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file_name: Option<String>,
}

/// How the frontend learns what this platform will actually do, instead of
/// discovering it from a failed command. Same posture as the Docker sandbox on
/// Windows: refuse clearly rather than silently do something weaker than the
/// user was promised.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct BrowserCapabilities {
    pub live_windows_supported: bool,
    pub reader_supported: bool,
    /// A stable token naming *why*, when something is off. The wording lives in
    /// the frontend's i18n.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub platform_note: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn json<T: Serialize>(value: &T) -> serde_json::Value {
        serde_json::to_value(value).expect("serialize")
    }

    /// The frontend's typed wrappers switch on these lowercase words; a
    /// renamed variant would fall through to "unknown" silently.
    #[test]
    fn tls_state_is_lowercase_and_defaults_to_unknown() {
        assert_eq!(json(&TlsState::Secure), "secure");
        assert_eq!(json(&TlsState::Insecure), "insecure");
        assert_eq!(json(&TlsState::Unknown), "unknown");
        assert_eq!(TlsState::default(), TlsState::Unknown);
        assert_eq!(
            serde_json::from_str::<TlsState>("\"secure\"").unwrap(),
            TlsState::Secure
        );
        assert!(serde_json::from_str::<TlsState>("\"Secure\"").is_err());
        assert_eq!(SecurityState::default().tls, TlsState::Unknown);
        assert!(!SecurityState::default().vpn_active);
    }

    /// The three-into-two mapping from the type doc: allow has no `reason`
    /// key at all, confirm is `allowed` *with* a reason, block is neither.
    #[test]
    fn url_verdict_encodes_allow_confirm_block_as_documented() {
        let base = UrlVerdict {
            allowed: true,
            reason: None,
            display_url: "https://example.com/".into(),
            punycode_warning: None,
            scheme: "https".into(),
            is_loopback: false,
        };
        let allow = json(&base);
        assert_eq!(allow["allowed"], true);
        assert!(allow.get("reason").is_none(), "allow carries no reason key");
        assert!(allow.get("punycode_warning").is_none());

        let confirm = json(&UrlVerdict {
            reason: Some("loopback".into()),
            is_loopback: true,
            ..base.clone()
        });
        assert_eq!(confirm["allowed"], true);
        assert_eq!(confirm["reason"], "loopback");

        let block = json(&UrlVerdict {
            allowed: false,
            reason: Some("scheme:file".into()),
            ..base
        });
        assert_eq!(block["allowed"], false);
        assert_eq!(block["reason"], "scheme:file");
    }

    /// The punycode warning is present exactly when the ASCII host differs —
    /// absent (not null) otherwise, so the chrome's "render both" rule keys on
    /// presence.
    #[test]
    fn punycode_warning_is_absent_not_null() {
        let plain = SecurityState {
            tls: TlsState::Secure,
            scheme: "https".into(),
            host_display: "example.com".into(),
            punycode_warning: None,
            vpn_active: false,
        };
        assert!(json(&plain).get("punycode_warning").is_none());
        let homograph = SecurityState {
            host_display: "аpple.com".into(),
            punycode_warning: Some("xn--pple-43d.com".into()),
            ..plain
        };
        let out = json(&homograph);
        assert_eq!(out["punycode_warning"], "xn--pple-43d.com");
        let back: SecurityState = serde_json::from_value(out).unwrap();
        assert_eq!(back, homograph);
    }

    /// The module's one rule: no type here carries a path. A download outcome
    /// is a flag plus a display name, and its default is "nothing saved".
    #[test]
    fn download_types_carry_a_display_name_and_never_a_path() {
        let outcome = DownloadOutcome::default();
        assert!(!outcome.saved);
        assert_eq!(json(&outcome), serde_json::json!({ "saved": false }));

        let request = DownloadRequest {
            download_id: "d1".into(),
            file_name: "setup.exe".into(),
            mime_type: None,
            size_bytes: Some(1024),
            sniff_mismatch: true,
        };
        let out = json(&request);
        let keys: Vec<&str> = out.as_object().unwrap().keys().map(String::as_str).collect();
        assert!(
            keys.iter().all(|k| !k.contains("path") && !k.contains("dir")),
            "a path-shaped key leaked: {keys:?}"
        );
        assert!(out.get("mime_type").is_none());
        assert_eq!(out["size_bytes"], 1024);
        assert_eq!(out["sniff_mismatch"], true);
    }

    /// `browser:blocked` names the live window it happened in when it has one,
    /// and omits the key — rather than writing null — when it does not.
    #[test]
    fn blocked_navigation_window_label_is_optional_on_the_wire() {
        let attributed = BlockedNavigation {
            display_url: "http://10.0.0.1/".into(),
            reason: "private-network".into(),
            window_label: Some("browser-3".into()),
        };
        assert_eq!(json(&attributed)["window_label"], "browser-3");
        let background = BlockedNavigation {
            window_label: None,
            ..attributed
        };
        assert!(json(&background).get("window_label").is_none());
        let back: BlockedNavigation =
            serde_json::from_str(r#"{"display_url":"u","reason":"no-host"}"#).unwrap();
        assert!(back.window_label.is_none());
    }

    /// A reader page round-trips with its nested security readout intact and
    /// keeps `requested_url` and `final_url` as distinct fields.
    #[test]
    fn reader_page_round_trips_with_both_urls() {
        let page = ReaderPage {
            requested_url: "http://example.com".into(),
            final_url: "https://example.com/".into(),
            display_url: "https://example.com/".into(),
            title: "Example".into(),
            html: "<p>hi</p>".into(),
            security: SecurityState {
                tls: TlsState::Secure,
                scheme: "https".into(),
                host_display: "example.com".into(),
                punycode_warning: None,
                vpn_active: true,
            },
            truncated: false,
            blocked_remote_assets: 2,
        };
        let back: ReaderPage = serde_json::from_value(json(&page)).unwrap();
        assert_eq!(back, page);
        assert_ne!(back.requested_url, back.final_url);
        let caps = BrowserCapabilities {
            live_windows_supported: false,
            reader_supported: true,
            platform_note: None,
        };
        assert!(json(&caps).get("platform_note").is_none());
    }
}
